File: checkIfPangram.py
# Approach 1- Array as counter (using count)
def checkIfPangram(sentence): # TC O(n) // SC O(1)
	freq=[0]*26

	for char in sentence:
	    #print(ord(char)-ord('a'))
	    freq[ord(char)-ord('a')]+=1

	for count in freq:
	    if count<1:
	        return False 

	return True

# Approach 2- Finding each letter
def checkIfPangram1(sentence): # TC O(n) // SC O(1)
	# We iterate over 'sentence' for 26 times, one for each letter 'curr_char'.
	for i in range(26):
	    curr_char = chr(ord('a') + i)

	    # If 'sentence' doesn't contain 'curr_char', it is not a pangram.
	    if sentence.find(curr_char) == -1:
	        return False

	# If we manage to find all 26 letters, it is a pangram.
	return True

# Approach 3- Using set 
def checkIfPangram2(sentence): # TC O(n) // SC O(1)
	# Add every letter of 'sentence' to hash set 'seen'.
    seen = set(sentence)
    
    # If the size of 'seen' is 26, then 'sentence' is a pangram.
    return len(seen) == 26

# Approach 4- Array as counter (using boolean)
def checkIfPangram3(sentence): # TC O(n) // SC O(1)
    # Array 'seen' of size 26.
    seen = [False] * 26

    # For every letter 'currChar', we find its ASCII code, 
    # and update value at the mapped index as true.
    for curr_char in sentence:
        seen[ord(curr_char) - ord('a')] = True
    
    # Once we finish iterating, check if 'seen' contains false.
    for status in seen:
        if not status:
            return False
    return True

# Approach 5- Bit Manipulation
def checkIfPangram4(sentence): # TC O(n) // SC O(1)
    # Initialize seen = 0 since we start with no letter.
    seen = 0
    
    # Iterate over 'sentence'.
    for curr_char in sentence:
        # Map each 'curr_char' to its index using its ASCII code.
        mapped_index = ord(curr_char) - ord('a')

        # 'curr_bit' represents the bit for 'curr_char'.
        curr_bit = 1 << mapped_index 

        # Use 'OR' operation since we only add its bit for once.
        seen |= curr_bit
    
    # Once we finish iterating, check if 'seen' contains 26 bits of 1.
    return seen == (1 << 26) - 1

print(checkIfPangram("thequickbrownfoxjumpsoverthelazydog"))
print(checkIfPangram1("thequickbrownfoxjumpsoverthelazydog"))
print(checkIfPangram2("thequickbrownfoxjumpsoverthelazydog"))
print(checkIfPangram3("thequickbrownfoxjumpsoverthelazydog"))
print(checkIfPangram4("thequickbrownfoxjumpsoverthelazydog"))

File: checkIfPangram.java
import java.util.*;
// Approach 1- Array as counter (using boolean)
class Solution{ // TC O(n) SC O(1)
	public boolean checkIfPangram(String sentence){
		boolean[] seen= new boolean[26];
		for(char c: sentence.toCharArray()){
			seen[c-'a']=true;
		}

		for(boolean status:seen){
			if(!status)
				return false;
		}
		return true;
	}
}

// Approach 2- Finding each letter
class Solution1{ // TC O(n) SC O(1)
    public boolean checkIfPangram(String sentence) {
        // We iterate over 'sentence' for 26 times, one for each letter 'currChar'.
        for (int i = 0; i < 26; ++i) {
            char currChar = (char)('a' + i);

            // If 'sentence' doesn't contain currChar, it is not a pangram.
            if (sentence.indexOf(currChar) == -1)
                return false;
        }
        
        // If we manage to find all 26 letters, it is a pangram.
        return true;
    }
}

// Approach 3- Using set 
class Solution2{ // TC O(n) SC O(1)
    public boolean checkIfPangram(String sentence) {
        // Empty set 'seen'.
        Set<Character> seen = new HashSet<>();
        
        // Iterate over 'sentence' add every letter to 'seen'.
        for (char currChar : sentence.toCharArray())
            seen.add(currChar);
        
        // If the size of 'seen' is 26, then 'sentence' is a pangram.
        return seen.size() == 26;
    }
}

// Coded by me 
// Apporach 4- Array as counter (using count)
class Solution3{ // TC O(n) SC O(1)
    public boolean checkIfPangram(String sentence) {
        int[] freq= new int[26];
        for (char c : sentence.toCharArray()){
            freq[c-'a']++;
            //System.out.println(c-'a');
        }
        
        for (int count: freq){
            if(count<1)
                return false;
        }
        
        return true;
    }
}

// Approach 5- Bit Manipulation
class Solution4{ // TC O(n) SC O(1)
    public boolean checkIfPangram(String sentence) {
        // Initialize seen = 0 since we start with no letter.
        int seen = 0;
        
        // Iterate over 'sentence'.
        for (char currChar : sentence.toCharArray()) {
            // Map each 'currChar' to its index using its ASCII code.
            int mappedIndex = currChar - 'a';

            // 'currBit' represents the bit for 'currChar'.
            int currBit = 1 << mappedIndex;

            // Use 'OR' operation since we only add its bit for once.
            seen |= currBit;
        }
        
        // Once we finish iterating, check if 'seen' contains 26 bits of 1.
        return seen == (1 << 26) - 1;
    }
}


class checkIfPangram{
  public static void main(String[] args){
    String sentence="thequickbrownfoxjumpsoverthelazydog";
    Solution s = new Solution();
    System.out.println(s.checkIfPangram(sentence));
    Solution1 s1 = new Solution1();
    System.out.println(s1.checkIfPangram(sentence));
    Solution2 s2 = new Solution2();
    System.out.println(s2.checkIfPangram(sentence));
    Solution3 s3= new Solution3();
    System.out.println(s3.checkIfPangram(sentence));
    Solution4 s4 = new Solution4();
    System.out.println(s4.checkIfPangram(sentence));
    
  }
}